ownCloud/Client/ClientQueryViewController.swift

```diff
@@ -645,6 +651,87 @@ class ClientQueryViewController: UITableViewController, Themeable {
 		}
 
 	}
+
+	func upload(itemURL: URL, name: String, viewDidAppearHandler: ClientActionVieDidAppearHandler? = nil, completionHandler: ClientActionCompletionHandler? = nil) {
+		if let progress = core.importFileNamed(name, at: query.rootItem, from: itemURL, isSecurityScoped: false, options: nil, placeholderCompletionHandler: nil, resultHandler: { [weak self](error, _ core, _ item, _) in
+			if error != nil {
+				Log.debug("Error uploading \(Log.mask(name)) file to \(Log.mask(self?.query.rootItem.path))")
+				completionHandler?(false)
+			} else {
+				Log.debug("Success uploading \(Log.mask(name)) file to \(Log.mask(self?.query.rootItem.path))")
+				completionHandler?(true)
+			}
+		}) {
+			self.progressSummarizer?.startTracking(progress: progress)
+		}
+	}
+
+	// MARK: - Navigation Bar Actions
+	@objc func uploadsBarButtonPressed(_ sender: UIBarButtonItem) {
+
+		let controller = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+
+		let photoLibrary = UIAlertAction(title: "Upload from your photo library".localized, style: .default, handler: { (_) in
+
+			func presentImageGalleryPicker() {
+				let picker = UIImagePickerController.regularImagePicker(with: .photoLibrary)
+				picker.delegate = self
+				OnMainThread {
+					self.present(picker, animated: true)
+				}
+			}
+
+			let permisson = PHPhotoLibrary.authorizationStatus()
+			switch permisson {
+
+			case .authorized:
+				presentImageGalleryPicker()
+			case .notDetermined:
+				PHPhotoLibrary.requestAuthorization({ newStatus in
+					if newStatus == .authorized {
+						presentImageGalleryPicker()
+					}
+				})
+
+			default:
+				PHPhotoLibrary.requestAuthorization({ newStatus in
+
+					if newStatus == .denied {
+						let alert = UIAlertController(title: "Missing permissions".localized, message: "This permission is needed to upload photos and videos from your photo library.".localized, preferredStyle: .alert)
+
+						let settingAction = UIAlertAction(title: "Settings".localized, style: .default, handler: { _ in
+							UIApplication.shared.open(URL(string: UIApplicationOpenSettingsURLString)!, options: [:], completionHandler: nil)
+						})
+						let notNowAction = UIAlertAction(title: "Not now".localized, style: .cancel)
+
+						alert.addAction(settingAction)
+						alert.addAction(notNowAction)
+
+						OnMainThread {
+							self.present(alert, animated: true)
+						}
+					}
+				})
+			}
+		})
+
+		let locationsDirectory = UIAlertAction(title: "Upload file".localized, style: .default) { _ in
+			let documentDirectory = UIDocumentPickerViewController(documentTypes: [kUTTypeData as String], in: .import)
```

Comment:
I think `documentDirectory` is confusing name here, since it is actually referring to a view controller instance

Reply:
I'll change it to `documentPickerVC`

Reply:
@pablocarmu I think we have put a rule into our coding guide, that abbreviations like `VC` shall not be used. Probably better to just name it `documentPickerViewController`. cc: @felix-schwarz

Reply:
changed